pkg/signalid/media.go

```diff
@@ -0,0 +1,229 @@
+// mautrix-signal - A Matrix-Signal puppeting bridge.
+// Copyright (C) 2025 Tulir Asokan
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+package signalid
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"io"
+
+	"maunium.net/go/mautrix/bridgev2/networkid"
+)
+
+type directMediaType byte
+
+const (
+	directMediaTypeAttachment directMediaType = iota
+	directMediaTypeGroupAvatar
+	directMediaTypeProfileAvatar
+)
+
+type DirectMediaInfo interface {
+	AsMediaID() (networkid.MediaID, error)
+}
+
+var (
+	_ DirectMediaInfo = (*DirectMediaAttachment)(nil)
+	_ DirectMediaInfo = (*DirectMediaGroupAvatar)(nil)
+	_ DirectMediaInfo = (*DirectMediaProfileAvatar)(nil)
+)
+
+type DirectMediaAttachment struct {
+	CDNID     uint64
+	CDNKey    string
+	CDNNumber uint32
+	Key       []byte
+	Digest    []byte
+	Size      uint32
+}
+
+func (m DirectMediaAttachment) AsMediaID() (mediaID networkid.MediaID, err error) {
+	buf := &bytes.Buffer{}
+
+	if err = binary.Write(buf, binary.BigEndian, directMediaTypeAttachment); err != nil {
+		return
+	} else if err = writeUvarint(buf, m.CDNID); err != nil {
+		return
+	} else if err = writeByteSlice(buf, []byte(m.CDNKey)); err != nil {
+		return
+	} else if err = writeUvarint(buf, uint64(m.CDNNumber)); err != nil {
+		return
+	} else if err = writeByteSlice(buf, m.Key); err != nil {
+		return
+	} else if err = writeByteSlice(buf, m.Digest); err != nil {
+		return
+	} else if err = writeUvarint(buf, uint64(m.Size)); err != nil {
+		return
+	}
+
+	return networkid.MediaID(buf.Bytes()), nil
+}
+
+type DirectMediaGroupAvatar struct {
+	UserID          []byte
+	GroupID         []byte
```

Comment:
Might make sense to validate lengths already when parsing user/group ids and use the proper fixed length types here (like `uuid.UUID` is a `[16]byte`)